test/mechanics/integrands/HistoryData.cpp

```diff
@@ -0,0 +1,397 @@
+#include "BoostUnitTest.h"
+
+#include "base/Group.h"
+#include "mechanics/cell/Cell.h"
+#include "mechanics/cell/CellInterface.h"
+#include "mechanics/cell/SimpleAssember.h"
+#include "mechanics/constitutive/laws/LinearElastic.h"
+#include "mechanics/constitutive/laws/MechanicsInterface.h"
+#include "mechanics/constraintsPde/Constraints.h"
+#include "mechanics/constraintsPde/ConstraintCompanion.h"
+#include "mechanics/dofs/DofNumbering.h"
+#include "mechanics/integrands/MomentumBalance.h"
+#include "mechanics/integrationtypes/IntegrationTypeTensorProduct.h"
+#include "mechanics/interpolation/InterpolationTrussLinear.h"
+#include "mechanics/mesh/MeshFem.h"
+#include "mechanics/mesh/MeshFemDofConvert.h"
+#include "mechanics/nodes/NodeSimple.h"
+
+#include <cassert>
+#include <functional>
+#include <map>
+#include <vector>
+
+using namespace Eigen;
+using namespace NuTo;
+using namespace NuTo::Groups;
+
+
+constexpr double SpecimenLength = 10.;
+constexpr unsigned int numElements = 4;
+
+//! @brief Standard history data management object.
+//! @tparam T: Data type of the history data
+template <typename T>
+class HistoryData
+{
+    std::map<int, std::vector<T>> mHisData;
+
+public:
+    //! @brief Returns the history data for a specific integration point
+    //! @param cellNum: Number of the cell containing the integration point
+    //! @param ipNum: Number of the integration point
+    //! @return History data of the integration point
+    const T& GetIPHistoryData(int cellNum, int ipNum) const
+    {
+        auto ipVecIt = mHisData.find(cellNum);
+        if (ipVecIt == mHisData.end())
+            throw Exception(__PRETTY_FUNCTION__, "Cell data not found! - Did you initialize the history data?");
+        return (*ipVecIt).second[ipNum];
+    }
+
+    //! @brief Returns the history data for a specific integration point
+    //! @param cellNum: Number of the cell containing the integration point
+    //! @param ipNum: Number of the integration point
+    //! @return History data of the integration point
+    T& GetIPHistoryData(int cellNum, int ipNum)
+    {
+        auto ipVecIt = mHisData.find(cellNum);
+        if (ipVecIt == mHisData.end())
+            throw Exception(__PRETTY_FUNCTION__, "Cell data not found! - Did you initialize the history data?");
+        return (*ipVecIt).second[ipNum];
+    }
+
+    //! @brief Initializes the history data
+    //! @param cells: Cellgroup that is related to the integrand
+    //! @param dof: Dof type
+    void InitializeHistoryData(const Groups::Group<CellInterface>& cells, DofType dof)
```

Comment:
I do not like the coupling to `Group`, `Cell` and `DofType` here. This is supposed to be some kind of a 2D container of `T`s. Why do we need those? Sure, for initialization. You can provide the sizes in a `map<int, int>` or something. To what value is everything initialized?

Reply:
We can also just use a single int for the size, if we can decide on a basic data type like a std::vector. The conditions for that are mentioned already. 
With the map<int,int> and the current underlying container the user needs to loop over all cells to get their ids and number of IPs.